packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart

```diff
@@ -1,17 +1,349 @@
+      return Padding(
+        padding: const EdgeInsets.only(bottom: 8.0),
+        child: Container(
+          decoration: BoxDecoration(
+            border: Border.all(color: Theme.of(context).focusColor),
+          ),
+          child: _controller.timelineMode == TimelineMode.frameBased
+              ? _FrameBasedTimelineFlameChart(
+                  _controller.frameBasedTimeline.data.selectedFrame,
+                  width: constraints.maxWidth,
+                  height: math.max(
+                    constraints.maxHeight,
+                    _frameBasedTimelineChartHeight(),
+                  ),
+                  selectionProvider: () =>
+                      _controller.frameBasedTimeline.data.selectedEvent,
+                  onSelection: (e) => _controller.selectTimelineEvent(e),
+                )
+              // TODO(kenz): implement full timeline flame chart.
+              : Container(
+                  color: Colors.black26,
+                  child: const Center(
+                    child: Text('TODO Full Timeline Flame Chart'),
+                  ),
+                ),
+        ),
+      );
+    });
+  }
+
+  double _frameBasedTimelineChartHeight() {
+    return (_controller.frameBasedTimeline.data.displayDepth + 2) *
+            rowHeightWithPadding +
+        sectionSpacing;
+  }
+}
+
+// TODO(kenz): Abstract core flame chart logic for use in other flame charts.
+class _FrameBasedTimelineFlameChart extends StatefulWidget {
+  _FrameBasedTimelineFlameChart(
+    this.data, {
+    @required this.height,
+    @required double width,
+    @required this.selectionProvider,
+    @required this.onSelection,
+  })  : duration = data.time.duration,
+        startInset = sideInset,
+        totalStartingWidth = width;
+
+  final TimelineFrame data;
+
+  final Duration duration;
+
+  final double startInset;
+
+  final double totalStartingWidth;
+
+  final double height;
+
+  final TimelineEvent Function() selectionProvider;
+
+  final VoidFunctionWithArg<TimelineEvent> onSelection;
+
+  double get startingContentWidth =>
+      totalStartingWidth - startInset - sideInset;
+
+  @override
+  _FrameBasedTimelineFlameChartState createState() =>
+      _FrameBasedTimelineFlameChartState();
+}
+
+class _FrameBasedTimelineFlameChartState
+    extends State<_FrameBasedTimelineFlameChart> {
+  static const startingScrollPosition = 0.0;
+  ScrollController _scrollControllerX;
+  ScrollController _scrollControllerY;
+  double scrollOffsetX = startingScrollPosition;
+  double scrollOffsetY = startingScrollPosition;
+
+  final List<FlameChartRow> rows = [];
+
+  TimelineController _controller;
+
+  StreamSubscription _selectedEventSubscription;
+
+  int get gpuSectionStartRow => widget.data.uiEventFlow.depth;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _controller = Controllers.of(context).timeline;
+    _selectedEventSubscription?.cancel();
```

Comment:
use the mixin that handles managing subscriptions rather than manually cancelling.